server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2ErrorStrategy.java

```diff
@@ -94,16 +98,58 @@ protected void reportMissingToken(Parser recognizer) {
     if (inErrorRecoveryMode(recognizer)) {
       return;
     }
+
     beginErrorCondition(recognizer);
-    String msg =
-        messageService.getMessage(
-            REPORT_MISSING_TOKEN,
-            errorMessageHelper.getExpectedText(recognizer),
-            ErrorMessageHelper.getRule(recognizer));
+    String msg = recognizer.getExpectedTokens().contains(Db2SqlLexer.END_EXEC)
+              ? messageService.getMessage(REPORT_MISSING_END_EXEC)
+              : messageService.getMessage(
+                      REPORT_MISSING_TOKEN,
+                      errorMessageHelper.getExpectedText(recognizer),
+                      ErrorMessageHelper.getRule(recognizer));
     recognizer.notifyErrorListeners(recognizer.getCurrentToken(), msg, null);
   }
 
   private String getOffendingToken(InputMismatchException e) {
     return getTokenErrorDisplay(e.getOffendingToken());
   }
+
+  @Override
+  public Token recoverInline(Parser recognizer) throws RecognitionException {
+    ParserRuleContext context = recognizer.getContext();
+    if (context instanceof Db2SqlParser.ExecRuleContext) {
+      int startIdx = context.start.getTokenIndex();
+      TokenStream input = recognizer.getInputStream();
+      input.seek(startIdx + 1); // SKIP EXEC_SQL
+      Token t = input.LT(1);
+      Db2SqlParser.SqlCodeContext sqlCodeContext = ((Db2SqlParser.ExecRuleContext) context).sqlCode();
+      ofNullable(sqlCodeContext.children).ifPresent(List::clear);
+      while (!isInAriaA(t) && !isSqlStopChar(t)
+              && t.getType() != Db2SqlLexer.DOT_FS
+              && t.getType() != Token.EOF) {
+          input.consume();
+          sqlCodeContext.addChild(recognizer.createTerminalNode(sqlCodeContext, t));
+          sqlCodeContext.stop = t;
```

Comment:
Could you just remove tokens that are to be kept instead of removing all of them and recreating those to be kept?
(including the `;` below)